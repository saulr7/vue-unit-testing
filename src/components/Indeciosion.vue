<template>
  <div>
    <h1>Indecision</h1>
    <img :src="img" alt="bg" />
    <div class="bg-dark"></div>
    <div class="container">
      <input type="text" placeholder="Question" v-model="question" />
      <div v-show="isValidQuestion">
        <p>{{ question }}</p>
        <h1>{{ answer }}</h1>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: "Indecision",

  data() {
    return {
      question: "",
      answer: null,
      img : "https://via.placeholder.com/250",
      isValidQuestion : false
    };
  },
  methods: {
    async getAnswer() {
      try {
        
        this.anwser = "Loading..."
       const {answer, image} = await fetch("https://yesno.wtf/api").then((res)=> res.json())
       this.answer = answer
       this.img = image
       this.isValidQuestion = true
      } catch (error) {
        console.log(error)
        this.answer = "cannot get response"        
      }
    },
  },
  watch: {
    question(value) {

    this.isValidQuestion = false
    console.log({value})
    if (!value.includes("?")) return;
      this.getAnswer()
    },
  },
};
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>