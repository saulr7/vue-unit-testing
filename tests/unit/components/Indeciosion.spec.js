import Indeciosion from "@/components/Indeciosion.vue";
import { shallowMount } from "@vue/test-utils";

describe("Indeciosion Component", () => {
  let wrapper;
  let clgSpy;
  let getAnswerSpy;
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://yesno.wtf/assets/yes/2.gif",
        }),
    })
  );
  beforeEach(() => {
    wrapper = shallowMount(Indeciosion);
    clgSpy = jest.spyOn(console, "log");
    getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("should match Snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should not fire searchs", async () => {
    const input = wrapper.find("input");
    await input.setValue("hola mundo");
    expect(clgSpy).toHaveBeenCalledTimes(1);
    expect(getAnswerSpy).not.toHaveBeenCalled();
  });
  test("should fire getAnswer on ? symbol is enter", async () => {
    const input = wrapper.find("input");
    await input.setValue("hola mundo ?");
    expect(getAnswerSpy).toHaveBeenCalledTimes(1);
  });
  test("get answer", async () => {
    await wrapper.vm.getAnswer();
    const img = wrapper.find("img");
    expect(img.exists()).toBeTruthy();
  });
  test("get answer faild API", async () => {
    fetch.mockImplementationOnce(() => Promise.reject("cannot get response"));
    await wrapper.vm.getAnswer();
    expect(wrapper.vm.answer).toBe("cannot get response");
    //cannot get response
  });
});
