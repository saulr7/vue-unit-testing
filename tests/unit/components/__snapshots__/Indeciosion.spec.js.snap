// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indeciosion Component should match Snapshot 1`] = `
<div>
  <h1>Indecision</h1><img src="https://via.placeholder.com/250" alt="bg">
  <div class="bg-dark"></div>
  <div class="container"><input type="text" placeholder="Question">
    <div style="display: none;">
      <p></p>
      <h1></h1>
    </div>
  </div>
</div>
`;
